tao_compiler/mlir/disc/transforms/lower_tf.cc

```diff
@@ -628,12 +633,438 @@ LogicalResult ConvertDequantizeOp::dequantize(TF::DequantizeOp op,
+  // all spatial dimensions.
+  const int64_t filter_channels = getDimSize(filter_ty, num_spatial_dims);
+  // TensorFlow convolution op verifies that the number of input channels is
+  // divisible by the number of filter channels.
+  // For depthwise convolution the feature_group_count argument would be set
+  // to the input feature dimension.
+  const int64_t feature_group_count = input_channels / filter_channels;
+  auto feature_group_count_attr = b.getNamedAttr(
+      "feature_group_count", b.getI64IntegerAttr(feature_group_count));
+
+  auto batch_group_count_attr =
+      b.getNamedAttr("batch_group_count", b.getI64IntegerAttr(1));
+
+  params.padding = b.create<tensor::FromElementsOp>(
+      loc, RankedTensorType::get(2 * num_spatial_dims, b.getI32Type()),
+      paddings);
+
+  NamedAttribute attrs[] = {rhs_dilations_attr, window_strides_attr,
+                            dimension_numbers_attr, feature_group_count_attr,
+                            batch_group_count_attr};
+  params.config = b.getDictionaryAttr(attrs);
+
+  return success();
+}
+
+LogicalResult matchAndRwriterQuantizedConv2DWithBiasAndRequantizeOp(
+    Operation* op) {
+  if (op->getNumOperands() != 9 || op->getNumResults() != 3)
+    return op->emitError() << "mismatch # operands or # results\n";
+
+  Value input = op->getOperand(0);
+  Value filter = op->getOperand(1);
+  Value bias = op->getOperand(2);
+  Value minInput = op->getOperand(3);
+  Value maxInput = op->getOperand(4);
+  Value minFilter = op->getOperand(5);
+  Value maxFilter = op->getOperand(6);
+  Value minFreezedOutput = op->getOperand(7);
+  Value maxFreezedOutput = op->getOperand(8);
+  Value output = op->getResult(0);
+  Value minOutput = op->getResult(1);
+  Value maxOutput = op->getResult(2);
+
+  auto inputTy = input.getType().dyn_cast<RankedTensorType>();
+  auto filterTy = filter.getType().dyn_cast<RankedTensorType>();
+  auto biasTy = bias.getType().dyn_cast<RankedTensorType>();
+  auto minInputTy = minInput.getType().dyn_cast<RankedTensorType>();
+  auto maxInputTy = maxInput.getType().dyn_cast<RankedTensorType>();
+  auto minFilterTy = minFilter.getType().dyn_cast<RankedTensorType>();
+  auto maxFilterTy = maxFilter.getType().dyn_cast<RankedTensorType>();
+  auto minFreezedOutputTy =
+      minFreezedOutput.getType().dyn_cast<RankedTensorType>();
+  auto maxFreezedOutputTy =
+      maxFreezedOutput.getType().dyn_cast<RankedTensorType>();
+  auto outputTy = output.getType().dyn_cast<RankedTensorType>();
+  auto minOutputTy = minOutput.getType().dyn_cast<RankedTensorType>();
+  auto maxOutputTy = maxOutput.getType().dyn_cast<RankedTensorType>();
+
+  // only try to match the op has static rank.
+  if (!inputTy || !filterTy || !biasTy || !minInputTy || !maxInputTy ||
+      !minFilterTy || !maxFilterTy || !minFreezedOutputTy ||
+      !maxFreezedOutputTy || !outputTy || !minOutputTy || !maxOutputTy)
+    return success();
```

Comment:
Is it ok to return `success()` since we've got a `tf.QuantizedConv2DWithBiasAndRequantizeOp` but cound not lower it. If the the op could not be handled here, anywhere else?

Reply:
Return `success` here is ok. We actually have two tf2mhlo passes. The first pass is our implementation for some special op or op having special attributes. The second pass is the community tf2mhlo pass.

It's follow the best practice of rewriter-pattern. One pattern just try to rewriter the pattern it can handle. Following such practice, we can have multiple  rewriter-patterns for the same op, each handling part of the variants.

If there are remaining `tf.QuantizedConv2DWithBiasAndRequantizeOp` ops after the pass, following community tf2mhlo pass will try to rewrite it. If the community pass fail as well, the whole pipeline will fail, with clear error message.